python/cudf/cudf/core/dataframe.py

```diff
@@ -341,27 +341,73 @@ def _setitem_tuple_arg(self, key, value):
+                    columns_df[col][scatter_map] = (
+                        value._data[col]
+                        if col in value_column_names
+                        else cudf.NA
+                    )
+
+            else:
+                value = cupy.asarray(value)
+                if cupy.ndim(value) == 2:
+                    # If the inner dimension is 1, it's broadcastable to
+                    # all columns of the dataframe.
+                    indexed_shape = columns_df.loc[key[0]].shape
+                    if value.shape[1] == 1:
+                        if value.shape[0] != indexed_shape[0]:
+                            raise ValueError(
+                                error_msg.format(
+                                    value1=value.shape,
+                                    value2=indexed_shape,
+                                )
+                            )
+                        for i, col in enumerate(columns_df._column_names):
+                            self._frame[col].loc[key[0]] = value[:, 0]
                     else:
-                        raise ValueError(
-                            f"shape mismatch: value array of shape "
-                            f"{value_df.shape} could not be "
-                            f"broadcast to indexing result of shape "
-                            f"{columns_df.shape}"
-                        )
+                        if value.shape != indexed_shape:
+                            raise ValueError(
+                                error_msg.format(
+                                    value1=value.shape,
+                                    value2=indexed_shape,
+                                )
+                            )
+                        for i, col in enumerate(columns_df._column_names):
+                            self._frame[col].loc[key[0]] = value[:, i]
                 else:
-                    value_cols = value_df._data.columns
-                for i, col in enumerate(columns_df._column_names):
-                    self._frame[col].loc[key[0]] = value_cols[i]
-            else:
-                for col in columns_df._column_names:
-                    self._frame[col].loc[key[0]] = value
+                    # handle cases where value is 1d object:
+                    # If the value is a range object or if the column axis
+                    # key is 0d, the indexed object is a series; therefore
+                    # the 1d array is assigned to the series along the column.
+                    if is_scalar(key[1]):
+                        for col in columns_df._column_names:
+                            self._frame[col].loc[key[0]] = value
+                    # Otherwise, there are two situations. The row axis key
+                    # is 0d or the key is a 2d indexer. In either of the
+                    # situation, the ith element in value corresponds to
+                    # the ith column in the indexed object.
```

Comment:
```suggestion
                    # Otherwise, there are two situations. The key on row axis can be
                    # a scalar or 1d. In either of the situation, the ith element in
                    # value corresponds to the ith row in the indexed object. If the key
                    # is 1d, a broadcast will happen.
```